File: day-04/part-2.ts
import { loadInput, sum } from "../utils";


const toPairAssignment = (row: string) => {
  const [a, x] = row.split(",");
  const first = a.split("-").map(Number);
  const second = x.split("-").map(Number);
  return [first, second];
};

const createSections = (a: number, b: number) => Array.from({ length: b - a + 1 }, (_, i) => a + i);

const getOverlaping = (ass: number[][][]) => ass.map(([[a, b], [x, y]]) => {
  const first = createSections(a, b);
  const second = createSections(x, y);

  const isOverlaping = first.some(f => second.includes(f));

  return isOverlaping ? 1 : 0;
});

const input = loadInput("input");
const rows = input.split("\n");

const assignments = rows.map(toPairAssignment);
const overlaps = getOverlaping(assignments);

console.log(sum(overlaps));

File: kotlin/src/main/kotlin/day04/part-2.kt
package day04.part2

import loadInput

fun main() {
    val input = loadInput("./src/main/kotlin/day04/input")
    val rows = input.split("\n")

    val overlaps = rows.map {
        val (first, second) = toSectionPair(it)
        val (a, b) = first
        val (x, y) = second
        if (a in x..y || b in x..y || x in a..b || y in a..b) 1 else 0
    }

    println(overlaps.sum())
}

fun toSectionPair(row: String) = row.split(",").map { it.split("-").map { it.toInt() } }
